### ISSUE TITLE
VS Code Extension: Redwood Outline is empty

### ISSUE THREAD

**Issue by andrewlamyw** [redwoodjs/redwood#1261]
## Maintainer Edit Oct. 14, 2020
> WORKAROUND: In most cases, simply "opening a file" (any file in the project) will fix the issue. 
>
> Check #1355 for more info

----

## Original Post
I have just downloaded Redwood IDE on my Visual Studio Code but the Redwood Outline section is empty. 
I have tried to restart the IDE but it does not resolve the issue. 
Do I need to do anything to activate it or is it a bug? 

![image](https://user-images.githubusercontent.com/11419166/94990675-8a715c00-05b0-11eb-9566-5e3eef8e5f8e.png)
![image](https://user-images.githubusercontent.com/11419166/94990863-a4f80500-05b1-11eb-80f7-90076330e6d9.png)
![image](https://user-images.githubusercontent.com/11419166/94991190-fd300680-05b3-11eb-93f2-a1611f84a133.png)

cc @aldonline @thedavidprice

**Comment by thedavidprice** [comment]
Thank you @andrewlamyw for taking the Extension for a spin! Handling over to Aldo.

**Comment by KrisCoulson** [comment]
Will second this I ran the same exact steps as @andrewlamyw outline continues to be empty

**Comment by aldonline** [comment]
Hi @andrewlamyw + @KrisCoulson,

Thanks for reporting!

Are you guys available for a quick Google Hangout session?
If yes, just join me here. I'll be online for the next few hours: https://meet.google.com/xdd-vkck-wkw.
I might no be in front of the computer at the moment. Just send me a chat and wait.

Two questions:
 
1. What happens if you open a `.js` file or the `redwood.toml` file in the editor? Does this fix the issue?
2. Can you take a look at the Output Channels and see if the Redwood Language Server Channel is present?

Just to give you guys a mental model of what's (possibly) going on. This is the lifecycle (get ready for something pretty ridiculous):

* VSCode starts
* It reads extension metadata for all extensions (their package.json) and finds that the Redwood IDE extension declares a "Redwood Outline" view. At this point it "might" draw an empty view. It still hasn't executed any extension code, this is all from a set of package.jsons being interpreted
* It starts the Extension Host, which is a separate process that will initialize extensions one by one
* The extension host activates the Redwood extension
* The extension runs a pre-activation step that installs dependencies (yarn...). This can take a moment (you'll see a status bar message with a spinner (ex: "redwoodjs-ide 0.0.9 installing dependencies..."). It also creates an output channel `decoupled.redwoodjs-ide-0.0.9 (preactivator)`
* At this point the "Redwood Outline" view should still be empty
* Once dependencies are ready, the extension starts executing code
* It creates an output channel called `decoupled.redwoodjs-ide-0.0.9`
* It will look for a redwood.toml in the workspace
* If it finds one, then it needs to start the Redwood Language Server in the backend (the one that actually parses the project, detects errors, etc)
* The Language Server is bundled with Redwood itself, so the extension looks for a file in node_modules:   `node_modules/@redwoodjs/structure/dist/language_server/start.js` 
* Then it starts the language server
* The language server creates an output channel called `Redwood Language Server`
* The Language Server starts processing the project. This can take seconds to a minute
* Once the first pass is ready, it will start talking back to the extension
* AT THIS POINT, THE OUTLINE APPEARS FOR THE FIRST TIME
* The language server is called at an interval to update diagnostics, the outline, etc. At any given point, the Language Server might also crash (because your project might be invalid, to a point where a basic assumption fails). If this happens, the extension will try to restart it at an interval

As you can see, there are many steps where things can go wrong.
The presence of output channels is a good place to start to figure out where things went wrong. Their are created in the following order:

1. `decoupled.redwoodjs-ide-0.0.9 (preactivator)`
2. `decoupled.redwoodjs-ide-0.0.9`
3. `Redwood Language Server`

I am adding instrumentation to the extension via Amplify, and adding more checks for invariants along the way, so I can get some visibility and catch these errors. But, in the meantime, if you feel adventurous

**Comment by aldonline** [comment]
@thedavidprice @mojombo reading the above you might get a glimpse of how much work was needed in the jamstackide just in the QA side of things. Imagine the same process, but for 5 frameworks + a bunch of other services (and multiple versions of each, on multiple platforms), and trying to work with Netlify Dev's super faulty detection heuristics without breaking it (to the point where we were tunneling things if necessary). Thank god for Sindre Sorhus' process management modules. We ended up realizing, in the end, that the best way to go was to build a little abstraction that was able query the "real state of things" and figure out if the user was already running dev servers, which ports were used by those services, or if she was in the middle of installing stuff, etc. So the state of the system was derived from the actual system, not an approximate model.
So, anyway, I can't help but find it odd when people tell me they they're building an extension or two as a side project

**Comment by aldonline** [comment]
I added a small fix in 0.0.10 and released

**Comment by andrewlamyw** [comment]
@thedavidprice Yeah, can't wait to use it! 😁

I have updated the Redwood IDE extension from version 0.0.9 to 0.0.13
![image](https://user-images.githubusercontent.com/11419166/95004024-8f69f600-0618-11eb-99c8-9ae6e4f57486.png)


@aldonline Thanks for your response😃. The following are the answers to your questions:

> What happens if you open a .js file or the redwood.toml file in the editor? Does this fix the issue?

> Can you take a look at the Output Channels and see if the Redwood Language Server Channel is present?

1. When I open a .js file, a pop up is shown with the message of `A request has failed. See the output for more information. Source: Redwood IDE (Extension)`. I click `Go to output` button, it opens up the output of Redwood Language Server but without any content.

1. When I open the `redwood.toml` file, nothing happen.

![image](https://user-images.githubusercontent.com/11419166/95003853-5cbefe00-0616-11eb-8ead-c30c09a4d7b8.png)
![image](https://user-images.githubusercontent.com/11419166/95003994-0eaafa00-0618-11eb-9555-d80ed8bff24c.png)

**Comment by aldonline** [comment]
Thanks @andrewlamyw!
It looks like the processPagesDir function fails when the pages dir doesn't exist. I added a fix here (https://github.com/redwoodjs/redwood/pull/1268). I can't release this quickly since it is a fix in Redwood itself, not the extension. I'll bring this up with @thedavidprice. Maybe we can do a patch release.
This is a good example of a class of error that is very typical in packages like @redwoodjs/structure, which should tolerate malformed projects without throwing but, at the same time, should not stray too far away from what's right.

**Comment by aldonline** [comment]
@andrewlamyw can you try with Redwood 0.19.2 and the latest version of the extension?

**Comment by thedavidprice** [comment]
@aldonline This is working well for me! BUT I've noticed the Outline does not show/trigger until I open a file. So if Open Editors is empty at the start, the Redwood Outline will be empty. But once the Outline displays, it stays displayed even if I close all files.

Thanks for all your great work here! I'd ❤️ making this official in v0.20.0 next week if you think it's ready.

**Comment by KrisCoulson** [comment]
@aldonline sorry I wasn't able to get back to you was super busy over the weekend. Since this post, I have updated to `0.19.2`

Still am getting nothing in the redwood outline and every time I open a JS file I treated this
![image](https://user-images.githubusercontent.com/6943688/95426934-9f752480-08fb-11eb-99bb-4a55d448bd61.png)

It then continues to keep pinging in the language server indefinitely 

![image](https://user-images.githubusercontent.com/6943688/95427071-d64b3a80-08fb-11eb-9bd9-dbb456c73cb9.png)

I am around daily if you want also just ping me on the discord channel for more feedback I would be happy to provide as much detail as I can

**Comment by AntonioMeireles** [comment]
@aldonline 

first let me congratulate you for your great and awesome work! 

now an UX nitpick - until #1280 is sorted could you consider to just ignor he output of `rw check` in the extension ? (or at least have a way to just ignore it ?) 

<img width="730" alt="image" src="https://user-images.githubusercontent.com/743527/95453502-e1858100-0962-11eb-9e2c-12e36c5428fc.png">

Thanks in advance, 

António

**Comment by aldonline** [comment]
Hi @AntonioMeireles!

Thanks for the kind words :) I really appreciate that you're using this and helping us getting it to a "useful" state!
I just opened a PR to fix the env visibility error. #1308 

I agree that we should have a way to turn on/off diagnostics.
Instead of reinventing the wheel we could copy what most linters do and allow turning errors on/off granularly. Maybe in redwood.toml?
I'll create an RFC issue so we can think this through ;)

**Comment by AntonioMeireles** [comment]
thank you us all ! this has the potential to be sort of game changer in a lot of ways ... :-) 

(thinking loud) regarding your idea, at sight, there's two potential angles to handle this ... treat it as a config / runtime option and instrument it via `redwood.toml`. or treat it as a plain linting  / idiomatic one and do it via ... the linter (adding options to perhaps `eslint-plugin-redwood`) . 

for testbeds both approaches would work and would be formally equivalent **but** the later one would scale a lot better imho for larger projects (specially ones involving multiple people) as one wouldn't want to plug idiomatic settings in the project's `redwood.toml` which should be immutable across team members. linters are plugable / with defaults easily overridable  (either via `eslintrc` or straight from `package.json`) and would allow end-user to select verbosity/noise level, etc...

All the best, 

António

**Comment by aldonline** [comment]
Hi @KrisCoulson ,

Looking at the stack trace it seems that the `RWRouter.filePath` property is unexpectedly null. There is only one place where the `RWRouter` gets instantiated. 

https://github.com/redwoodjs/redwood/blob/main/packages/structure/src/model/RWProject.ts#L113

```ts
  getRouter = () => {
    return new RWRouter(this.pathHelper.web.routes, this)
  }
```

I would love to know what's going on there (is `this.pathHelper.web.routes` actually null?, if yes, are there more errors in pathHelper?)

If you feel adventurous, you can help me figure this one out. For example, you can modify the compiled code directly to capture some logging.

`{your_project}/node_modules/@redwoodjs/structure/dist/model/RWProject.js`

Around line 63, you'll find something similar to this code:

```js
class RWProject extends _ide.BaseNode {
  constructor(opts) {
    super();
    this.opts = opts;
    this.parent = undefined;

    this.getRouter = () => {
      return new _RWRouter.RWRouter(this.pathHelper.web.routes, this);
    };
  }
```

You can add some logging here.
Unfortunately, if you just use `console.log` we won't be able to see that (at least not easily), but you can just write out to a file somewhere. (Since you're on a windows machine, you might need to pick a different filename to write to).

Ex:

```js
class RWProject extends _ide.BaseNode {
  constructor(opts) {
    super();
    this.opts = opts;
    this.parent = undefined;

    this.getRouter = () => {
      require("fs").writeFileSync("/tmp/rw-logging-output.json", JSON.stringify(this.pathHelper))
      return new _RWRouter.RWRouter(this.pathHelper.web.routes, this);
    };
  }
```

If you modify this code just reload the extension.

**Comment by andrewlamyw** [comment]
> @andrewlamyw can you try with Redwood 0.19.2 and the latest version of the extension?

@aldonline I'm using Redwood 0.19.2 and Redwood IDE v0.0.14 but still couldn't see the Redwood Outline.

![image](https://user-images.githubusercontent.com/11419166/95569573-5686b280-0a58-11eb-8a47-60ad0a578717.png)
![image](https://user-images.githubusercontent.com/11419166/95569847-b715ef80-0a58-11eb-9267-e720a27a05ca.png)

**Comment by aldonline** [comment]
@andrewlamyw + @KrisCoulson ,
I think I found the source of the error :). Working on a fix now.
There is a fs.existsSync() call somewhere deep in Redwood that fails in certain conditions on windows. Looks permission related. This results in the path for the router file being set to null.
Anyway. Just letting you know so you don't spend time trying to diagnose further.

**Comment by aldonline** [comment]
@andrewlamyw + @KrisCoulson, v0.19.3 fixes several windows issues. It is possible that it fixes the underlying cause of the issues you were experiencing.

**Comment by andrew-hwahin** [comment]
Thanks, @aldonline and @thedavidprice.

I'm having difficulty in upgrading @redwoodjs packages by running `yarn rw upgrade`. It is showing that all of my dependencies are up to date.

My project is currently using v0.19.2 but when I use my cursor to point at the @redwoodjs/core, the version is showing 0.19.1 instead of 0.19.3

![image](https://user-images.githubusercontent.com/61768800/95696562-37616e00-0c6e-11eb-9df7-1085e2d26813.png)

**Comment by thedavidprice** [comment]
@andrew-hwahin That's very strange indeed and clearly not working correctly. In order to test the VS Code Extension, you can manually upgrade the packages by changing each `@redwoodjs/*` package to `v0.19.3` in the following 3 files:
- `package.json`
- `web/package.json`
- `api/package.json`

Save, and then install with `yarn` command. 

Does that seem to do the trick?

**Comment by andrew-hwahin** [comment]
@thedavidprice I've manually upgraded the packages by following your instructions but the VS Code Extension still does not work. 
![image](https://user-images.githubusercontent.com/61768800/95708993-49ec9f00-0c90-11eb-929d-6350ced9a5d6.png)
![image](https://user-images.githubusercontent.com/61768800/95709092-80c2b500-0c90-11eb-974e-3afa32118688.png)

**Comment by thedavidprice** [comment]
@andrew-hwahin Ah, I see the confusion here. You're referencing the _installed_ package within the `node_modules/` directory. In general, you should never need to change or modify these installed packages directly. Consider it "read only" as it's managed by 1) your projects `package.json` settings and 2) yarn package manager.

Looking at my example Blog project repository, here are the `package.json` files that set the specific Redwood versions:
- Root directory: https://github.com/thedavidprice/redwood-tutorial-test/blob/master/package.json
   - change `"@redwoodjs/core": "^0.19.3"`
- Web/ directory: https://github.com/thedavidprice/redwood-tutorial-test/blob/master/web/package.json
    -  change `"@redwoodjs/auth": "^0.19.3", "@redwoodjs/forms": "^0.19.3", "@redwoodjs/router": "^0.19.3", "@redwoodjs/web": "^0.19.3",` (note: you might not have auth package installed)
- API/ directory: https://github.com/thedavidprice/redwood-tutorial-test/blob/master/api/package.json
    - change `"@redwoodjs/api": "^0.19.3"`

**Comment by andrew-hwahin** [comment]
@thedavidprice Thanks for the detailed explanation🙏. I changed the `package.json` file as you mentioned, in the Root, Web, and API directory. I took the screenshot in `node_modules` just to verify that the packages are updated successfully. 😀

Hence, in that case, upgrading RedwoodJS does not resolve the issue yet.

**Comment by thedavidprice** [comment]
@andrew-hwahin that's a bummer. Thanks again for testing and reporting back.

@Tobbe and @RobertBroersma  any chance either you have time to test the latest Redwood VS Code extension against Redwood v0.19.3?

**Comment by aldonline** [comment]
Hi @andrew-hwahin,

I can see from the screenshots that you hit a strange edge-case where the extension doesn't run the preinstallation step fully. I'm trying to replicate on my end.

In the meantime, can you try the following?

1. Uninstall and re-install the Redwood IDE extension
2. Create a new app from scratch using `yarn create redwood-app` and *then* open VSCode, what do you see?
Tip: There are several Output Channels that are of interest. They all have word "redwood" word in them.

**Comment by Tobbe** [comment]
I'm setting up a fresh RW app right now. Will install the IDE extension next. It will be the first time I install the extension

**Comment by Tobbe** [comment]
![image](https://user-images.githubusercontent.com/30793/95915860-1309ac80-0da8-11eb-8fe7-0ca36a2a4533.png)

It was empty. So I pressed the little 🔃 icon in the upper right corner. Didn't seem to do anything. So I opened up redwood.toml and now I got content in the outline. Not sure if it was just being slow, and it would have populated had I just waited, or if the above sequence of interactions triggered its appearance.

**Comment by thedavidprice** [comment]
^^ @aldonline FYI that Tobbe is running Windows

Thanks for the help here @Tobbe!

**Comment by KrisCoulson** [comment]
@aldonline  Just upgraded to 19.3 and looks like ` It is possible that it fixes the underlying cause of the issues you were experiencing.` was a good possibility it is working great now. This is definitely going to be a huge time saver for navigation! Awesome work!

**Comment by aldonline** [comment]
@Tobbe thanks for the help! Your description helped me identify a separate issue (https://github.com/redwoodjs/redwood/issues/1355)

@KrisCoulson Good to know! The Outline is still work in progress. Feel free to request new features ;). You can find the code that builds the outline here (https://github.com/redwoodjs/redwood/blob/main/packages/structure/src/outline/outline.ts) in case you want to contribute/fix something

**Comment by aldonline** [comment]
**TIP:** In most cases, simply "opening a file" (any file in the project) will fix the issue. Check #1355 for more info

**Comment by aldonline** [comment]
^^ @thedavidprice can we pin this comment to the top somehow? Or edit the issue description?

**Comment by thedavidprice** [comment]
@aldonline See edit in OP. Does that work?

**Comment by aldonline** [comment]
@thedavidprice yes, that's perfect! Thanks! (and thanks @Tobbe for figuring this one out)

**Comment by andrewlamyw** [comment]
Hi, I've just tried it out again with RedwoodJS v0.20 and Redwood IDE v0.0.16 and it's working fine now, thank you all very much.🙏♥
